spec/transaction_spec.cr

```diff
@@ -0,0 +1,158 @@
+
+    with_witness do |w|
+      with_dummy_connection do |cnn|
+        cnn.transaction do |tx|
+          if tx.is_a?(DummyDriver::DummyTransaction)
+            t = tx
+            w.check
+          end
+        end
+      end
+    end
+
+    t.committed.should be_true
+    t.rolledback.should be_false
+  end
+
+  it "transaction with block from connection should be rolledback if raise DB::Rollback" do
+    t = uninitialized DummyDriver::DummyTransaction
+
+    with_witness do |w|
+      with_dummy_connection do |cnn|
+        cnn.transaction do |tx|
+          if tx.is_a?(DummyDriver::DummyTransaction)
+            t = tx
+            w.check
+          end
+          raise DB::Rollback.new
+        end
+      end
+    end
+
+    t.rolledback.should be_true
+    t.committed.should be_false
+  end
+
+  it "transaction with block from connection should be rolledback if raise" do
+    t = uninitialized DummyDriver::DummyTransaction
+
+    with_witness do |w|
+      with_dummy_connection do |cnn|
+        expect_raises(FooException) do
+          cnn.transaction do |tx|
+            if tx.is_a?(DummyDriver::DummyTransaction)
+              t = tx
+              w.check
+            end
+            raise FooException.new
+          end
+        end
+      end
+    end
+
+    t.rolledback.should be_true
+    t.committed.should be_false
+  end
+
+  it "transaction can be committed within block" do
+    with_dummy_connection do |cnn|
+      cnn.transaction do |tx|
+        tx.commit
+      end
+    end
+  end
+
+  it "transaction can be rolledback within block" do
+    with_dummy_connection do |cnn|
+      cnn.transaction do |tx|
+        tx.rollback
+      end
+    end
+  end
+
+  it "transaction can't be committed twice" do
+    with_dummy_connection do |cnn|
+      cnn.transaction do |tx|
+        tx.commit
+        expect_raises(DB::Error, "Transaction already closed") do
+          tx.commit
+        end
+      end
+    end
+  end
+
+  it "transaction can't be rolledback twice" do
+    with_dummy_connection do |cnn|
+      cnn.transaction do |tx|
+        tx.rollback
+        expect_raises(DB::Error, "Transaction already closed") do
+          tx.rollback
+        end
+      end
+    end
+  end
+
+  it "return connection to pool after transaction block in db" do
+    DummyDriver::DummyConnection.clear_connections
+
+    with_dummy do |db|
+      db.transaction do |tx|
+        db.pool.is_available?(DummyDriver::DummyConnection.connections.first).should be_false
+      end
+      db.pool.is_available?(DummyDriver::DummyConnection.connections.first).should be_true
+    end
+  end
+
+  it "releasing result_set from within transaction should not return connection to pool" do
+    cnn = uninitialized DB::Connection
+    with_dummy do |db|
+      db.transaction do |tx|
+        cnn = tx.connection
+        cnn.scalar "1"
+        db.pool.is_available?(cnn).should be_false
+      end
+      db.pool.is_available?(cnn).should be_true
+    end
+  end
```

Comment:
I'd add another spec here: ensure that a connection is not returned to the pool if a nested transaction is commited/rolledback, but there is still an outer active one.